File: modules/src/jumpDetection.cc
#include <emscripten.h>
#include <emscripten/bind.h>
#include <string>
#include <vector>
#include <list>
#include <chrono> 
#include "json.hpp"

// for convenience
using json = nlohmann::json;
using namespace emscripten;

class Jump {
    public:
    Jump(){}
    std::string process(const std::string &str) {
        this->calculate_current_frame_distance(str);
        this->calculate_current_distance_mean();
        this->calculate_current_direction();
        std::string res = this->add_json(str);
        return res;
    }

    void calculate_current_distance_mean() {
        this->current_distance_mean = this->current_distance_mean * 0.8 + this->current_frame_distance * 0.2;
    }

    void calculate_current_direction() {
        this->current_speed_mean = this->current_speed_mean * 0.9  + abs(this->current_frame_distance - this->current_distance_mean) * 0.1;
        if(this->current_frame_distance - this->current_distance_mean > this->amp_threshold) {
            if(this->current_jump != 0) {
                if(this->frameShiftFilterCount > 2) {
                    this->shiftDirectionStepPerMinutes();
                    this->current_jump = 0;
                    this->frameShiftFilterCount = 0;
                } else {
                    this->frameShiftFilterCount = this->frameShiftFilterCount + 1;
                }
            } else {
                this->frameShiftFilterCount = 0;
            }        
        } 
        if(this->current_frame_distance - this->current_distance_mean < -this->amp_threshold) {
            if(this->current_jump != 1) {
                if(this->frameShiftFilterCount > 2) {
                    this->shiftDirectionStepPerMinutes();
                    this->current_jump = 1;
                    this->frameShiftFilterCount = 0;
                } else {
                    this->frameShiftFilterCount = this->frameShiftFilterCount + 1;
                }
            } else {
                this->frameShiftFilterCount = 0;
            }
        } 
        if(abs(this->current_frame_distance - this->current_distance_mean) < this->amp_threshold) {
            this->fpmStopCount = this->fpmStopCount + 1;
            if(this->fpmStopCount > 30) {
                this->current_jump = 0;
                this->fpm = 0;
            }
        } else {
            this->fpmStopCount = 0;
        }
        this->current_height = this->current_jump == 1? this->current_speed_mean:0;
    }

    void shiftDirectionStepPerMinutes() {
        long long now = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count();

        while (this->times.size() > 0 && this->times.front() <= now - 5000) {
            this->times.pop_front();
        }
        this->times.push_back(now);
        this->fpm = this->times.size() * 20;
    }

    void calculate_current_frame_distance(const std::string &str) {

        this->current_distance_mean = 0;
        auto j = json::parse(str);
        // get data from json
        float num1 = j["pose_landmark"]["keypoints"][23]["y"];
        float num2 = j["pose_landmark"]["keypoints"][24]["y"];
        float num3 = j["pose_landmark"]["keypoints"][25]["y"];
        float num4 = j["pose_landmark"]["keypoints"][26]["y"];
        float num5 = j["pose_landmark"]["keypoints"][27]["y"];
        float num6 = j["pose_landmark"]["keypoints"][28]["y"];
        float weighted_sum = num1 + num2 + num3 + num4 + num5 + num6;
        this->current_distance_mean = weighted_sum;
    }

    std::string add_json(const std::string &str) {

        auto j = json::parse(str);
        // get data from json
        j["action_detection"]["jump"]["up"] = this->current_jump;
        j["action_detection"]["jump"]["strength"] = this->current_height;
        std::string s = j.dump();
        return s;
    }


    private:
    float current_distance_mean = 0;
    float current_speed_mean = 0;
    float current_frame_distance = 0;
    float current_jump = 0;
    float current_height = 0;
    float speed_mean = 0;
    float amp_threshold = 0.25;
    bool monitor_process = false;
    float fpm = 0;
    std::list<long long> times;
    float fpmStopCount = 0;
    float frameShiftFilterCount = 0;
};

// Binding code
EMSCRIPTEN_BINDINGS(my_class_example) {
  class_<Jump>("Jump")
    .constructor<>()
    .function("process", &Jump::process)
    ;
}

File: wasm/app/src/action_detection/jump_detection.js
const {performance} = require('perf_hooks')

var jump_processor = {
    current_distance_mean : 0,
    current_speed_mean: 0,
    current_frame_distance : 0,
    current_jump: 0,
    current_height: 0,
    speed_mean: 0,
    amp_threshold: 0.25,
    monitor_process : false,
    fpm : 0,
    times : [],
    fpmStopCount : 0,
    frameShiftFilterCount : 0,
    resultData: {},

    process : function(pose, resultData, monitor = false){
        this.resultData = resultData
        this.calculate_current_frame_distance(pose)
        this.calculate_current_distance_mean(pose)
        this.calculate_current_direction(pose, monitor)
        this.monitor_process = monitor
        return this.resultData
    }, 

    /*1. calculate current frame point weighted summation
     */
    calculate_current_frame_distance: function(pose) {
        this.current_frame_distance = 0
        this.current_frame_distance = this.weighted_sum(pose)
    },

    calculate_current_distance_mean: function(pose) {
        this.current_distance_mean = this.current_distance_mean * 0.8 + this.current_frame_distance * 0.2
    },

    calculate_current_direction: function(pose, monitor = false) {
        this.current_speed_mean = this.current_speed_mean * 0.9  + Math.abs(this.current_frame_distance - this.current_distance_mean) * 0.1
        if(this.current_frame_distance - this.current_distance_mean > this.amp_threshold) {
            if(this.current_jump != 0) {
                if(this.frameShiftFilterCount > 2) {
                    this.shiftDirectionStepPerMinutes()
                    this.current_jump = 0 
                    this.frameShiftFilterCount = 0
                } else {
                    this.frameShiftFilterCount = this.frameShiftFilterCount + 1
                }
            } else {
                this.frameShiftFilterCount = 0
            }        
        } 
        if(this.current_frame_distance - this.current_distance_mean < -this.amp_threshold) {
            if(this.current_jump != 1) {
                if(this.frameShiftFilterCount > 2) {
                    this.shiftDirectionStepPerMinutes()
                    this.current_jump = 1 
                    this.frameShiftFilterCount = 0
                } else {
                    this.frameShiftFilterCount = this.frameShiftFilterCount + 1
                }
            } else {
                this.frameShiftFilterCount = 0
            }
        } 
        if(Math.abs(this.current_frame_distance - this.current_distance_mean) < this.amp_threshold) {
            this.fpmStopCount = this.fpmStopCount + 1 
            if(this.fpmStopCount > 30) {
                this.current_jump = 0
                this.fpm = 0
            }
        } else {
            this.fpmStopCount = 0
        }
        // console.log(this.current_jump)
        // console.log(this.current_speed_mean)
        this.current_height = this.current_jump == 1? this.current_speed_mean:0
        this.resultData.action_detection.jump = {
            "up" : this.current_jump,
            "strength" : this.current_height
        }
        if (monitor) {
            console.log("freq : " + this.fpm)
            console.log("strength : " + this.current_speed_mean)
            this.resultData.monitor = {
                "rawData": this.current_jump,
                "watchData" :this.current_frame_distance
            }
        }
    },

    shiftDirectionStepPerMinutes : function() { 
        const now = performance.now();
        while (this.times.length > 0 && this.times[0] <= now - 5000) {
            this.times.shift();
        }
        this.times.push(now);
        this.fpm = this.times.length * 20;
    },

    /**2. using a filter to determin summation mean in each frame  */

    weighted_sum: function(pose) {
        res = pose.keypoints[23].y + pose.keypoints[24].y + pose.keypoints[25].y + pose.keypoints[26].y + pose.keypoints[27].y + pose.keypoints[28].y
        return res
    }
}

module.exports = jump_processor